File: breadth-first-search/keys-and-rooms/keys-and-rooms.java
class Solution {
  public boolean canVisitAllRooms(List<List<Integer>> rooms) {
    int n = rooms.size();
    Queue<Integer> queue = new ArrayDeque<>();
    boolean[] seen = new boolean[n];

    queue.add(0);
    seen[0] = true;

    while (!queue.isEmpty()) {
      Integer num = queue.poll();
      List<Integer> room = rooms.get(num);

      for (int keys : room) {
        if (!seen[keys]) {
          queue.add(keys);
          seen[keys] = true;
        }
      }
    }

    for (boolean v : seen) {
      if (!v) {
        return false;
      }
    }
    return true;
  }
}

File: breadth-first-search/keys-and-rooms/keys-and-rooms.js
/**
 * @param {number[][]} rooms
 * @return {boolean}
 */
function canVisitAllRooms(rooms) {
  const n = rooms.length;
  const queue = [];
  const seen = new Array(n);

  queue.push(0);
  seen[0] = true;

  while (queue.length) {
    const num = queue.shift();
    const room = rooms[num];

    for (const key of room) {
      if (!seen[key]) {
        queue.push(key);
        seen[key] = true;
      }
    }
  }

  for (const v of seen) {
    if (!v) {
      return false;
    }
  }
  return true;
}
